awscli/customizations/history/db.py

```diff
@@ -0,0 +1,246 @@
+# Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import os
+import uuid
+import time
+import json
+import datetime
+import threading
+import logging
+from collections import MutableMapping
+
+from botocore.history import BaseHistoryHandler
+
+from awscli.compat import sqlite3
+from awscli.compat import ensure_text_type
+
+
+LOG = logging.getLogger('awscli.history.db')
+
+
+class DatabaseConnection(object):
+    _CREATE_TABLE = """
+    CREATE TABLE IF NOT EXISTS records (
+      id TEXT,
+      request_id TEXT,
+      source TEXT,
+      event_type TEXT,
+      timestamp INTEGER,
+      payload TEXT
+    )"""
+    _ENABLE_WAL = """PRAGMA journal_mode=WAL"""
+    _DATABASE_NAME = 'history.db'
+    _DEFAULT_DATABASE_FILENAME = os.path.expanduser(
+        os.path.join('~', '.aws', 'cli', 'history', _DATABASE_NAME))
+
+    def __init__(self, db_filename=None):
+        if db_filename:
+            self._db_filename = db_filename
+        elif os.environ.get('AWS_HISTORY_PATH'):
+            self._db_filename = os.path.join(
+                os.environ.get('AWS_HISTORY_PATH'), self._DATABASE_NAME)
+        else:
+            self._db_filename = self._DEFAULT_DATABASE_FILENAME
+
+        if self._db_filename != ':memory:' and \
+           not os.path.isdir(os.path.dirname(self._db_filename)):
+            os.makedirs(os.path.dirname(self._db_filename))
+        self._connection = sqlite3.connect(
+            self._db_filename, check_same_thread=False, isolation_level=None)
+        self._ensure_database_setup()
+
+    def execute(self, query, *parameters):
+        self._connection.execute(query, *parameters)
+
+    def cursor(self):
```

Comment:
We discussed removing the cursor didn't we?  I'd suggest simplifying this interface and not exposing cursors at all.  We shouldn't need it.

Reply:
For writing yes it uses the connection and autocommit mode. I do not see a way to read directly from the connection though, the __iter__ and all the fetch* commands are on the cursor only as far as I can tell.

Reply:
It's exactly the same as inserts: https://docs.python.org/2/library/sqlite3.html#using-shortcut-methods